icicle/backend/cpu/src/hash/cpu_merkle_tree.cpp

```diff
@@ -40,42 +40,41 @@ namespace icicle {
 
         if (0 < layer_idx) {
           // update nof_hashes to the next layer (below)
-          const uint64_t next_layer_total_size = layer_hashes[layer_idx - 1].output_size();
-          nof_hashes = nof_hashes * cur_layer.m_hash.default_input_chunk_size() / next_layer_total_size;
+          const uint64_t cur_layer_input_size = layer_hashes[layer_idx].default_input_chunk_size();
+          nof_hashes = nof_hashes * cur_layer_input_size / layer_hashes[layer_idx - 1].output_size();
 
           // Calculate path_size
-          m_pruned_path_size +=
-            layer_hashes[layer_idx].default_input_chunk_size() - layer_hashes[layer_idx - 1].output_size();
-          m_full_path_size += layer_hashes[layer_idx].default_input_chunk_size();
+          m_pruned_path_size += cur_layer_input_size - layer_hashes[layer_idx - 1].output_size();
+          m_full_path_size += cur_layer_input_size;
         }
       }
     }
 
-    // TODO: handle size
+    ~CPUMerkleTreeBackend()
+    {
+      for (auto& pair : m_map_segment_id_2_inputs) {
+        delete pair.second; // Ensure all dynamically allocated memory is freed
+      }
+    }
+
     eIcicleError build(const std::byte* leaves, uint64_t leaves_size, const MerkleTreeConfig& config) override
     {
       TasksManager<HashTask> task_manager(get_nof_workers(config)); // Run workers.
       if (m_tree_already_built) {
         ICICLE_LOG_ERROR << "Tree cannot be built more than one time";
         return eIcicleError::INVALID_ARGUMENT;
       }
-      const uint64_t expected_input_size = m_layers[0].m_nof_hashes * m_layers[0].m_hash.default_input_chunk_size();
-      if (leaves_size < expected_input_size) {
-        ICICLE_LOG_ERROR << "CPU Merkle tree: Expecting " << expected_input_size << " bytes in input, got "
-                         << leaves_size << ". Note: Padding is currently not supported but will be soon";
-        return eIcicleError::INVALID_ARGUMENT;
-      }
-      if (leaves_size > expected_input_size) {
-        ICICLE_LOG_ERROR << "CPU Merkle tree: Expecting " << expected_input_size << " bytes in input, got "
-                         << leaves_size << ". Leaves size cannot exceeds tree size.";
+      // build a vector with the leaves that needs padding.
+      std::vector<std::byte> padded_leaves;
+      if (!init_layers_db(config, leaves_size) || !init_padded_leaves(padded_leaves, leaves, leaves_size, config)) {
         return eIcicleError::INVALID_ARGUMENT;
```

Comment:
why did you remove the ICICLE_LOG_ERROR? every time you return an error you'd better log the error, so the user can understand how to fix his code

Reply:
thre error messages remain. They moved inside the functions init_layers_db, init_padded_leaves